modules/geolocation/geolocation.py

```diff
@@ -0,0 +1,330 @@
+        ground_points = []
+        for vec_down in vec_downs:
+            result, ground_point = self.__ground_intersection_from_vector(
+                vec_camera_position,
+                vec_down,
+            )
+            if not result:
+                return False, None
+
+            ground_points.append(ground_point)
+
+        # Get the image to ground mapping
+        src = np.array(self.__perspective_transform_sources, dtype=np.float32)
+        dst = np.array(ground_points, dtype=np.float32)
+        try:
+            # Pylint does not like the OpenCV module
+            # pylint: disable=no-member
+            matrix = cv2.getPerspectiveTransform(
+                src,
+                dst,
+            )
+            # pylint: enable=no-member
+        # All exceptions must be caught and logged as early as possible
+        # pylint: disable=bare-except
+        except:
+            # TODO: Logging
+            return False, None
+        # pylint: enable=bare-except
+
+        return True, matrix
+
+    @staticmethod
+    def __convert_detection_to_world_from_image(detection: detections_and_time.Detection,
+                                                perspective_transform_matrix: np.ndarray) \
+            -> "tuple[bool, detection_in_world.DetectionInWorld | None]":
+        """
+        Applies the transform matrix to the detection.
+        perspective_transform_matrix: Must be affine.
+        """
+        if not camera_properties.is_matrix_r3x3(perspective_transform_matrix):
+            return False, None
+
+        if not np.allclose(
+            perspective_transform_matrix[2, :],
+            np.array([0.0, 0.0, 1.0], dtype=np.float32),
+            atol=FLOAT_PRECISION_ABS_TOLERANCE,
+        ):
+            # Last row must be 0, 0, 1 for matrix to be affine
+            return False, None
+
+        centre = detection.get_centre()
+        top_left, top_right, bottom_left, bottom_right = detection.get_corners()
+
+        input_centre = np.array([centre[0], centre[1], 1.0], dtype=np.float32)
+        # More efficient to multiply a matrix than looping over the points
+        # Transpose to columns from rowss
+        input_vertices = np.array(
+            [
+                [top_left[0], top_left[1], 1.0],
+                [top_right[0], top_right[1], 1.0],
+                [bottom_left[0], bottom_left[1], 1.0],
+                [bottom_right[0], bottom_right[1], 1.0],
+            ],
+            dtype=np.float32,
+        ).T
+
+        # Single row/column does not need transpose
+        output_centre = perspective_transform_matrix @ input_centre
+        # Transpose back to rows from columns
+        output_vertices = (perspective_transform_matrix @ input_vertices).T
+
+        ground_centre = np.array(
+            [
+                output_centre[0] / output_centre[2],
+                output_centre[1] / output_centre[2],
+            ],
+            dtype=np.float32,
+        )
+
```

Comment:
This is the part of the code I don't really understand. Can you please explain why we have to divide by `output_centre[2]`

Reply:
This is a homogeneous/perspective divide (see: https://en.wikipedia.org/wiki/Transformation_matrix#Perspective_projection ). I can add a comment.